llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,430 @@
+  SmallVector<MCPhysReg, 16>
+  getPriorityReplacementRegs(const Register Reg,
+                             const SmallSet<MCPhysReg, 16> &ForbiddenPhysRegs);
+
+  bool isWAWCandidate(const Register &Reg, const MachineInstr &MI,
+                      const SmallVector<MCPhysReg, 16> &UsedPhyRegs,
+                      const SmallVector<MCPhysReg, 16> &UsedVirtRegs);
+
+  bool isIdentityCopy(const MachineInstr &MI);
+
+  // Check if the UnDef is the last define for a Register, or if there is
+  // another Def/Undef that will later complete the Definition of the
+  // Register. If so, it is too early to determine, if the Register is a
+  // candidate for a WAW replacement, or not. This is especially interesting,
+  // if Subregisters are used to hold a constant and this shared
+  // subregister-constant is equal for multiple Registers. A renamig here
+  // would destroy the shared usage of the sub-register and force the
+  // insertion of redundant Instructions.
+  bool isUndefReserved(
+      const MachineOperand &Def, const MachineInstr &MI,
+      const MachineBasicBlock &MBB,
+      std::map<unsigned, SmallVector<const MachineInstr *, 4>> &UndefPartReuse);
+
+  bool isLastDef(const MachineOperand &Def, const MachineInstr &DefInstr,
+                 const MachineBasicBlock &MBB);
+
+  bool reuseSubRegs(SmallVector<const MachineInstr *, 4> &OrigInstr,
+                    SmallVector<const MachineInstr *, 4> &OtherInstr);
+
+  bool containEqualSubRegDefs(const MachineInstr *InstrA,
+                              const MachineInstr *InstrB);
+
+  void addAliasRegs(SmallSet<MCPhysReg, 16> &ForbiddenPhysRegs,
+                    const MCPhysReg &PhysReg);
+
+  Register getAssignedPhysReg(const MachineOperand &MO) const;
+};
+
+Register AIEWawRegRewriter::getAssignedPhysReg(const MachineOperand &MO) const {
+  Register PhysReg = MO.getReg();
+  if (PhysReg.isVirtual())
+    PhysReg = VRM->getPhys(PhysReg);
+  return PhysReg;
+}
+
+bool AIEWawRegRewriter::runOnMachineFunction(MachineFunction &MF) {
+
+  MachineLoopInfo &MLI = getAnalysis<MachineLoopInfo>();
+
+  SmallVector<const MachineBasicBlock *, 4> LoopMBBs =
+      AIELoopUtils::getInnermostLoopBlocks(MLI);
+
+  if (LoopMBBs.empty())
+    return false;
+
+  this->MF = &MF;
+  MRI = &MF.getRegInfo();
+  TRI = static_cast<const AIEBaseRegisterInfo *>(MRI->getTargetRegisterInfo());
+  VRM = &getAnalysis<VirtRegMap>();
+  LRM = &getAnalysis<LiveRegMatrix>();
+  LIS = &getAnalysis<LiveIntervals>();
+  Modified = false;
+
+  LLVM_DEBUG(dbgs() << "*** WAW Loop Register Rewriting: " << MF.getName());
+  LLVM_DEBUG(dbgs() << " ***\n");
+
+  for (const MachineBasicBlock *MBB : LoopMBBs)
+    renameMBBPhysRegs(MBB);
+
+  return Modified;
+}
+
+bool AIEWawRegRewriter::renameMBBPhysRegs(const MachineBasicBlock *MBB) {
+  bool Modified = false;
+  LLVM_DEBUG(dbgs() << "WAW Reg Renaming BasicBlock "; MBB->dump();
+             dbgs() << "\n");
+  SmallVector<MCPhysReg, 16> UsedPhyRegs;
```

Comment:
nit: I feel that `UsedPhyRegs` and  `UsedVirtRegs`  are good candidates for a bitset.

Reply:
fixed with a Bitvector